Consolidate windows functions into rasterio.windows

perrygeo opened the pull request:
Moving windows functions out of `__init__.py`,  `_base.pyx` and `_io.pyx` into a unified windows module.

The Dataset methods are implemented as a `rasterio.io.WindowMethodsMixin` which can be inherited by subclasses to provide the same windows methods as wrappers around the `rasterio.windows` functions.

resolves #810 and goes a long ways towards #803

perrygeo commented:
@sgillies @brendan-ward  this is ready for review.

perrygeo commented:
@sgillies the is ready for your review. 

@brendan-ward you may be interested in this Mixin strategy for the purposes of the wrapping ndarray, crs and transform into a class, #798

brendan-ward commented on rasterio/windows.py:
```diff
@@ -88,5 +140,172 @@ def intersect(*windows):
+    return x_min, y_min, x_max, y_max
+
+
+def crop_window(window, height, width):
```
@perrygeo now that this and other functions above and below are namespaced in `rasterio.windows` consider dropping `window_*` and `*_window` from the function names.

perrygeo replied:
For the naming of the `rasterio.windows.window` function - should we leave as is or change to something like `from_bounds`?

brendan-ward commented on rasterio/windows.py:
```diff
@@ -88,5 +140,172 @@ def intersect(*windows):
+    return (r_start, r_stop), (c_start, c_stop)
+
+
+def get_index(x, y, affine, op=math.floor, precision=6):
```
@perrygeo not sure that this belongs here, it doesn't strike me as a window specific operation.

perrygeo replied:
it's a utility function required by a few of the windows functions. I'll make it a private `_get_index`

brendan-ward commented on rasterio/windows.py:
```diff
@@ -88,5 +140,172 @@ def intersect(*windows):
+    return row, col
+
+
+def get_window(left, bottom, right, top, affine, precision=6):
```
@perrygeo `affine` should be `transform` now.

brendan-ward replied:
@perrygeo consider changing function name, this isn't particularly descriptive here.  Maybe `from_bounds` or `window_from_bounds`?  Or possibly even just `window`?

perrygeo replied:
I think it's probably ok to drop `get_window` and inline it with `window` (which is the only function that calls it - no need for redundancy)

brendan-ward commented on tests/test_windows.py:
```diff
@@ -0,0 +1,175 @@
+    with rasterio.open('tests/data/RGB.byte.tif') as src:
+        left, bottom, right, top = src.bounds
+        dx, dy = src.res
+        eps = 1.0e-8
```
@perrygeo consider making `eps` a global.

sgillies replied:
@perrygeo: semi-random thought here... this could be a great place to try Hypothesis, vary `eps` to make sure our algorithm doesn't burst into flames.

perrygeo replied:
started #825 to track the ideas around property-based testing

brendan-ward commented on tests/test_windows.py:
```diff
@@ -0,0 +1,175 @@
+from rasterio.windows import (
+    window, window_bounds, window_transform, eval_window, window_index, window_shape)
+
+def test_window_method():
```
@perrygeo I think this test function should focus on testing the functions defined in `rasterio.windows`; those that become methods via the mixin should probably be in a different test module focused on testing the `RasterReader` class.

brendan-ward commented on tests/test_windows.py:
```diff
@@ -0,0 +1,175 @@
+
+
+def test_eval_window_bad_structure():
+    with pytest.raises(ValueError):
```
@perrygeo here and next function, could possibly simplify down to 
```
bad_params = [
    ((1, 2, 3), 10, 10),
    ((1, 2), 10, 10),
    (((1, 0), 2), 10, 10)
]
for params in bad_params:
    with pytest.raises(ValueError):
        eval_window(*params)
```

brendan-ward commented on tests/test_windows.py:
```diff
@@ -0,0 +1,175 @@
+    with pytest.raises(ValueError):
+        window_shape(((10, 20), (10, None)))
+
+def test_window_shape_errors2():
```
@perrygeo why not combine with prior function?

brendan-ward commented on tests/test_windows.py:
```diff
@@ -0,0 +1,175 @@
+    assert window_shape(((~0, None), (~0, None)), 100, 90) == (1, 1)
+    assert window_shape(((None, ~0), (None, ~0)), 100, 90) == (99, 89)
+
+def test_eval():
```
@perrygeo can this get combined with `test_eval_window` above?

brendan-ward commented:
@perrygeo this is looking good, definitely a win to have more tests.  Mixin looks like a good move here.  👍 

Please consider adding numpy style docstrings, esp. for the functions now in `rasterio.windows`.

sgillies commented:
I like this, too. I'll second @brendan-ward's suggestion of removing `window` from function names.

perrygeo commented:
As soon as I started renaming the `rasterio.windows` functions, I realized that we were relying on those names being imported into `rasterio._base` and other modules imported from there. So it's a bit more involved to make the change - but well worth it I think. I'll ping everyone again for round 2 of review when it's in good shape again...

perrygeo commented:
Because `get_index` is more of a transform operation (the inverse of @geowurster's `rasterio.transform.xy` function in #805) I moved it into the transform operation to set the stage for pulling index, xy and ul into another io mixin class.

perrygeo commented:
@sgillies @brendan-ward round 2 ready for review. Since last round, I

* cleaned up the tests a bit per feedback above 
* removed the `windows_` from the function names
* removed the redudant `get_window` function in favor of `from_bounds` (formerly window)
* changed the `get_index` function to use affine multiplication operator rather than gdal-style math.
* moved `get_index` to `rasterio.transform`

perrygeo commented:
Note: still need to add numpy-style docstrings but otherwise should be good to go

brendan-ward commented on tests/test_windows.py:
```diff
@@ -4,132 +4,99 @@
+        evaluate(window, 10, 10)
 
 
 def test_eval_window_invalid_dims():
```
@perrygeo you could probably use the above pytest parameterize annotation approach here too

brendan-ward commented:
Other than minor comment above, this is looking good 👌

sgillies commented on rasterio/io.py:
```diff
@@ -7,17 +7,50 @@
+    `rasterio.windows` module.
+
+    A subclass with this mixin MUST provide the following
+    properties: `transform`, `height` and `width`
```
👍
